### DISCUSSION TITLE
Shouldn't positional properties of records be get-only instead of init-only?

### DISCUSSION THREAD

**Discussion by thomaslevesque** [dotnet/csharplang#4282] [General] [answered]
If I declare a record like this:
```csharp
public record Foo(string Name);
```

Then the following code is legal:
```csharp
var f = new Foo("Hello") { Name = "WAT?" };
```

The `Name` property is specified twice: as a constructor argument, and using object initializer syntax. This seems wrong, and I don't think this should be allowed. The `Name` property should be get-only, not init-only.

Sure, it's not a very big issue, I could just "not do this". But it's annoying, because when you have both positional properties and "normal" properties, Intellisense shows properties that have already been initialized in the constructor. For instance, if I add a non-positional property to my previous example:
```csharp
public record Foo(string Name)
{
    public string Description { get; init; }
}
```
Then intellisense shows this (this is in LinqPad, but Visual Studio exhibits the same behavior):

![image](https://user-images.githubusercontent.com/1864281/103280367-85198700-49d0-11eb-8a09-0c58ad309ceb.png)

I realize it's probably too late to fix that, since C# 9 has shipped, but I wonder if this is something that had been considered during the design phase?

If it can't be fixed in the language, maybe VS could exclude positional properties when suggesting properties for object initializer syntax (the same way it excludes already initialized properties).

What do you think?

**Comment by HaloFour** [comment, marked as answer]
Positional records have `init` property accessors to support `with`:

```cs
record Person(string First, string Last);

var person1 = new Person("John", "Smith");
var person2 = person1 with { Last = "Jones" };

// compiles into
var person2 = person1.<>Clone();
person2.Last = "Jones";
```

**Reply by thomaslevesque** [reply]
Ah, yes, that makes sense. Still, the Intellisense issue is annoying, but I guess that's not really a problem in the language itself.
